python/perf-kernels/flash-attention.py

```diff
@@ -127,6 +164,19 @@ def _attn_fwd_inner(
         m_ij = tl.maximum(m_i, tl.max(qk, 1))
         qk = qk - m_ij[:, None]
         p = tl.math.exp2(qk)
+        # CAVEAT: Must update l_ij before applying dropout
+        l_ij = tl.sum(p, 1)
+        # Note about the conflicts of Flash attention algorithm and PyTorch's CUDA implementation
+        # PyTorch needs to return softmax(qk) (dropout mask encoded in sign bits)
+        # While Flash attention paper computer the dropout AFTER exp2(qk- m_ij)
```

Comment:
> While Flash attention paper computer the dropout AFTER exp2(qk- m_ij)

I'm a bit confused by this. The CUDA impl also applies dropout after exp2(qk-l_ij).

[Here](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/transformers/cuda/flash_attn/flash_fwd_kernel.h#L408) is where exp2(qk - l_ij) is computed.

And [here](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/transformers/cuda/flash_attn/flash_fwd_kernel.h#L430) is where dropout is applied, after.

Reply:
The mask we get out of this is much easier and simpler to use in testing.  The testing was already adjusted.  The CUDA implementation differs from the paper, which is what is referenced here.

Reply:
I think what I'm getting at is that the mask is applied at the wrong location. It needs to be applied after doing the softmax, which would be after the exponential.

Notice that dropout sets some values to 0.0f, and others are scaled by 1/(1-dropout_p). You do this scaling outside the loop on the PV result, which is correct *only* if the dropout is applied post exponential. Otherwise you must scale by the exp(1/(1-dropout_p).